### PR TITLE
[Component]: Dialog

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Fixes https://github.com/brave/leo/issues/71

@petemill what do you think of https://github.com/brave/leo/pull/260/files#diff-4a25d419101f95ef68e52670ca1834d529de65c0f2d2b216ddc3a0cab6136b15R6-R17, which automatically creates a container for Dialog popups, rather than you needing to add one to the dom?

If you think it's a good idea I can go back to the alert centre and use the same pattern there.

@aguscruiz there are a few areas where I think I need your input on defaults.
1. Should dialogs have a close button by default (currently no)
2. Should dialogs have a back button by default (currently no)
3. Should dialogs be closable with the escape key by default (currently yes)
4. Should dialogs be closable by clicking outside by default (currently yes)

### CODE REVIEW

File: src/components/dialog/dialog.svelte [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    background: var(--leo-color-container-background);
+    padding: var(--leo-spacing-32) var(--leo-spacing-32) var(--leo-spacing-40)
+      var(--leo-spacing-32);
+  }
+
+  .leo-dialog.mobile {
+    width: 374px;
+  }
+
+  .leo-dialog:not(.modal) {
```
Comment (petemill) [inline src/components/dialog/dialog.svelte:118]: Doesn't it always have the class `modal`?

File: src/components/dialog/dialog.svelte [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  }
+
+  :global .leo-dialog .actions ::slotted(*),
+  :global .leo-dialog div[slot='actions'] {
+    display: flex;
+    flex-direction: row;
+    align-items: center;
+    justify-content: stretch;
+    gap: var(--leo-spacing-16);
+  }
```
Comment (petemill) [inline src/components/dialog/dialog.svelte:170]: Perhaps a comment here explaining what these `:global` commands are attempting to do?

File: src/components/dialog/dialog.svelte [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  on:click|capture={(e) => {
+    if (!dialog) return
+
+    const rect = dialog.getBoundingClientRect()
+    const clickedOutside =
+      e.clientX < rect.x ||
+      e.clientY < rect.y ||
+      e.clientX > rect.x + rect.width ||
+      e.clientY > rect.y + rect.height
+    if (isOpen && clickedOutside && backdropClickCloses) {
```
Comment (petemill) [inline src/components/dialog/dialog.svelte:90]: how about early return if these conditions aren't met? That way we can avoid the `getBoundingClientRect` call

File: src/components/dialog/dialogHelpers.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+export const confirm = (message: string, options?: ConfirmDialogOptions) =>
+  dialog({
+    title: message,
+    actions: [
+      {
+        result: false,
+        kind: 'outline',
+        text: options?.cancelText ?? 'Cancel'
```
Comment (petemill) [inline src/components/dialog/dialogHelpers.ts:84]: I believe(?) this is the first time we're introducing some default text in to a Leo component, so we should probably think about internationalisation. Perhaps a provider-based approach where consuming apps can provide the strings that all Leo components need, like:
```ts
type LeoStrings = { cancel: string, ok: string }

Leo.ProvideStrings = function(strings: LeoStrings)
```

Not sure if we can get away with sending the current locale only or send all locales and also tell leo which is the current locale...

File: src/components/dialog/dialogHelpers.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    title: options.title,
+    body: options.body,
+    actions: options.actions,
+    resolve: (result?: T) => {
+      dialogs.update((d) => d.filter((i) => i !== info))
+      resolve(result)
+    }
+  }
+
+  dialogs.update((d) => [...d, info])
```
Comment (petemill) [inline src/components/dialog/dialogHelpers.ts:53]: Technically this part could be done inside the Promise callback, could it not, avoiding both the `resolve` assignment and the possible but extremely unlikely possibility that resolve is called before it's assigned. Although I'm not sure when the Promise callback is called so it _might_ actually be impossible that happens...

**Review by petemill — APPROVED** [review]
Some interesting techniques - and sure why not offer both ways to show a dialog if it works well!

File: src/components/dialog/dialog.svelte (hunk 1 above)
Comment (fallaciousreasoning) [reply src/components/dialog/dialog.svelte:118]: Wow, so it does read like that.

I'd originally written
`class:modal={modal}` but when I format the file it gets changed to `class:modal`, which I guess is shorthand (it still seems to work).

File: src/components/dialog/dialogHelpers.ts (hunk 5 above)
Comment (fallaciousreasoning) [reply src/components/dialog/dialogHelpers.ts:53]: I think the earliest a promise can resolve is in the next tick, so it shouldn't ever happen. That said, yours reads nicer :laughing:

File: src/components/dialog/dialogHelpers.ts (hunk 4 above)
Comment (fallaciousreasoning) [reply src/components/dialog/dialogHelpers.ts:84]: I agree, we should definitely think about this. For now though, I'm going to kick the can down the road and force you to provide strings.

**Pull request merged**